README.md

[![Arduino CI](https://github.com/RobTillaart/DS28CM00/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/DS28CM00/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/DS28CM00/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/DS28CM00/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/DS28CM00/actions/workflows/jsoncheck.yml)
[![GitHub issues](https://img.shields.io/github/issues/RobTillaart/DS28CM00.svg)](https://github.com/RobTillaart/DS28CM00/issues)

[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/DS28CM00/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/DS28CM00.svg?maxAge=3600)](https://github.com/RobTillaart/DS28CM00/releases)
[![PlatformIO Registry](https://badges.registry.platformio.org/packages/robtillaart/library/DS28CM00.svg)](https://registry.platformio.org/libraries/robtillaart/DS28CM00)


# DS28CM00

Arduino library for I2C DS28CM00 unique identification chip.  \[48 bit\]


## Description

THe DS28CM00 IC has a 64 bit address consisting of one type byte, 0x50, 6 bytes unique serial and a CRC
over the previous 7 bytes. This results in an 48 bits unique ID giving 281.474.976.710.656 combinations.
If that number is not unique enough, you could use 2 or more of them with an I2C multiplexer.

Since version 0.3.0 the library supports reading only the first n bytes from the serial.
Sometimes less bytes are sufficient, and less bytes takes less time.

The DS28CM00 IC has the same addressing scheme as the better known DS18B20, except for the type byte.
Therefore one could use an DS18B20 as an unique ID chip in a similar way and getting a temperature sensor
as bonus. 

The DS28CM00 can work in 2 modes, I2C and SMBus mode. Check datasheet for details.


#### Related

- https://github.com/RobTillaart/UUID


## Interface

```cpp
#include "DS28CM00.h"
```

- **DS28CM00(TwoWire \*wire = &Wire)** Constructor, with default Wire as I2C bus.
The device has a fixed address of 0x50.
- **bool begin()** initializes library, sets default DS28CM00_I2C_MODE.
Returns false if mode cannot be set or if device cannot be found.
- **bool isConnected()** returns true if device can be found on I2C bus.
- **bool getUID(uint8_t \* buffer, uint8_t size = 8)** copy unique serial number into buffer.
Size is default 8, maximum size, size = 0 returns false. 
Less unique bytes can be enough sometimes.
Returns false if action failed.
- **bool setI2CMode()** set DS28CM00_I2C_MODE (0x00). Returns false if action failed.
- **bool setSMBusMode()** set DS28CM00_SMBUS_MODE (0x01). Returns false if action failed.
- **bool getMode(uint8_t &mode)** returns mode set above. Returns false if action failed.


|  value  |  mode                   |
|:-------:|:-----------------------:|
|  0x00   |  DS28CM00_I2C_MODE      |
|  0x01   |  DS28CM00_SMBUS_MODE    |
|  0xFF   |  DS28CM00_MODE_UNKNOWN  |


## Operation

See examples.

The class is not tested extensively e.g. with ESP32 / ESP8266.
An example sketch for ESP32 exist and compiles. Feedback welcome.


## Future

#### Must

- test with ESP32 and other platforms. (feedback welcome).

#### Should

- only extend on request / bugs
- measure performance I2C 400 500 600 kHz ...

#### Could

- cache the serial? (costs 8 bytes) + one flag.
  - If (cached filled) => faster!

#### Wont

- **uint8_t getError()** add error handling? - I2C; #bytes; mode; ...(too much for simple device)
- **bool setMode()** public? no.

## Support

If you appreciate my libraries, you can support the development and maintenance.
Improve the quality of the libraries by providing issues and Pull Requests, or
donate through PayPal or GitHub sponsors.

Thank you,



CHANGELOG.md
# Change Log DS28CM00

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.3.1] - 2023-12-11
- refactor API, begin()
- update readme.md


## [0.3.0] - 2023-09-21
- add Wire1 support for ESP32
- refactored ESP32 interface, no constructor => but a begin()
- add parameter size to **bool getUID(uint8_t \* buffer, uint8_t size = 8)**
- add example **DS28CM00_getUID.ino**
- renamed some internal constants for readability

----

## [0.2.5] - 2022-11.02
- add changelog.md
- add rp2040 to build-CI
- update readme.md

## [0.2.4] - 2021-12-17
- update library.json
- update license
- add isConnected()
- minor edits

## [0.2.3] - 2021-10-26
- update build-CI
- add default Wire in constructor

## [0.2.2] - 2020-12-20
- add Arduino-CI + unit test

## [0.2.1] - 2020-06-07
- fix library.json

## [0.2.0] - 2020-04-11
- refactor
- add #pragma once
- multiple Wire
- ESP support (start)

----

## [0.1.0] - 2017-07-15-  initial version

## [0.1.00]
- initial version






test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2020-12-20
// PURPOSE: unit tests for the DS28CM00
//          https://github.com/RobTillaart/DS28CM00
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// https://github.com/Arduino-CI/arduino_ci/blob/master/cpp/unittest/Assertion.h#L33-L42
// ----------------------------
// assertEqual(expected, actual)
// assertNotEqual(expected, actual)
// assertLess(expected, actual)
// assertMore(expected, actual)
// assertLessOrEqual(expected, actual)
// assertMoreOrEqual(expected, actual)
// assertTrue(actual)
// assertFalse(actual)
// assertNull(actual)
// assertNotNull(actual)

#include <ArduinoUnitTests.h>

#include "Wire.h"
#include "DS28CM00.h"


unittest_setup()
{
}

unittest_teardown()
{
}


unittest(test_constructor_I)
{
  uint8_t   uid[8];
  DS28CM00  DS28(&Wire);

  DS28.begin();
  uint8_t mode;

  assertFalse(DS28.getMode(mode));           // not connected...
  assertEqual(DS28CM00_MODE_UNKNOWN, mode);

  assertTrue(DS28.setSMBusMode());           // apparently 
  assertFalse(DS28.getMode(mode));
  assertEqual(DS28CM00_MODE_UNKNOWN, mode);

  assertTrue(DS28.setI2CMode());
  assertFalse(DS28.getMode(mode));
  assertEqual(DS28CM00_MODE_UNKNOWN, mode);
}


unittest(test_constructor_II)
{
  uint8_t   uid[8];
  DS28CM00  DS28;     // use default Wire

  DS28.begin();
  uint8_t mode;

  assertFalse(DS28.getMode(mode));           // not connected...
  assertEqual(DS28CM00_MODE_UNKNOWN, mode);
}

unittest_main()

// --------


DS28CM00.cpp
//
//    FILE: DS28CM00.cpp
//  AUTHOR: Rob Tillaart
// PURPOSE: Library for the DS28CM00 unique identification chip.
// VERSION: 0.3.1
//     URL: https://github.com/RobTillaart/DS28CM00


#include "DS28CM00.h"


#define DS28CM00_DEVICE_ADDRESS      0x50
#define DS28CM00_UID_REGISTER        0x00
#define DS28CM00_CONTROL_REGISTER    0x08


DS28CM00::DS28CM00(TwoWire *wire)
{
  _wire = wire;
}


bool DS28CM00::begin()
{
  if (!isConnected()) return false;
  return setI2CMode();
}


bool DS28CM00::isConnected()
{
  _wire->beginTransmission(DS28CM00_DEVICE_ADDRESS);
  int rv = _wire->endTransmission();
  return rv == 0;
}


bool DS28CM00::getMode(uint8_t &mode)
{
  mode = DS28CM00_MODE_UNKNOWN;
  _wire->beginTransmission(DS28CM00_DEVICE_ADDRESS);
  _wire->write(DS28CM00_CONTROL_REGISTER);
  int rv = _wire->endTransmission();
  if (rv != 0) return false;

  uint8_t read = _wire->requestFrom(DS28CM00_DEVICE_ADDRESS, 1);
  if (read < 1) return false;

  mode = _wire->read();
  return true;
}


bool DS28CM00::getUID(uint8_t * buffer, uint8_t size)
{
  _wire->beginTransmission(DS28CM00_DEVICE_ADDRESS);
  _wire->write(DS28CM00_UID_REGISTER);
  int rv = _wire->endTransmission();
  if (rv != 0) return false;

  if (size == 0) return false;
  if (size > 8) size = 8;
  uint8_t read = _wire->requestFrom((uint8_t)DS28CM00_DEVICE_ADDRESS, size);
  if (read < size) return false;

  for (uint8_t i = 0; i < size; i++) 
  {
    buffer[i] = _wire->read();
  }

  return true;
}


////////////////////////////////////////////////////
//
//  PRIVATE
//
bool DS28CM00::setMode(uint8_t mode)
{
  _wire->beginTransmission(DS28CM00_DEVICE_ADDRESS);
  _wire->write(DS28CM00_CONTROL_REGISTER);
  _wire->write(mode);
  int rv = _wire->endTransmission();
  return rv == 0;
}



//  -- END OF FILE --



DS28CM00.h
#pragma once
//
//    FILE: DS28CM00.h
//  AUTHOR: Rob Tillaart
// PURPOSE: Library for the DS28CM00 unique identification chip.
// VERSION: 0.3.1
// HISTORY: See DS28CM00.cpp
//     URL: https://github.com/RobTillaart/DS28CM00


#include "Arduino.h"
#include "Wire.h"


#define DS28CM00_LIB_VERSION            (F("0.3.1"))
#define DS28CM00_I2C_MODE               0x00
#define DS28CM00_SMBUS_MODE             0x01
#define DS28CM00_MODE_UNKNOWN           0xFF


class DS28CM00
{
public:
  explicit DS28CM00(TwoWire *wire = &Wire);

  bool     begin();                //  default DS28CM00_I2C_MODE
  bool     isConnected();

  bool     getUID(uint8_t * buffer, uint8_t size = 8);

  bool     setI2CMode()           { return setMode(DS28CM00_I2C_MODE); };
  bool     setSMBusMode()         { return setMode(DS28CM00_SMBUS_MODE); };
  bool     getMode(uint8_t &mode);

private:
  bool     setMode(uint8_t mode);
  TwoWire* _wire;
};


//  -- END OF FILE --

